aten/src/ATen/native/eigen/SparseBlas.cpp

```diff
@@ -0,0 +1,272 @@
+
+namespace {
+
+void inline csr_indices_to_result_dtype_inplace(
+    const c10::ScalarType& dtype,
+    const at::Tensor& input) {
+  if (input.layout() == kSparseCsr) {
+    static_cast<at::SparseCsrTensorImpl*>(input.unsafeGetTensorImpl())
+        ->set_member_tensors(
+            input.crow_indices().to(dtype),
+            input.col_indices().to(dtype),
+            input.values(),
+            input.sizes());
+  }
+}
+
+void inline csr_col_indices_and_values_resize(
+    const at::Tensor& input,
+    int64_t nnz) {
+  if (input.layout() == kSparseCsr) {
+    static_cast<SparseCsrTensorImpl*>(input.unsafeGetTensorImpl())
+        ->set_member_tensors(
+            input.crow_indices(),
+            input.col_indices().resize_({nnz}),
+            input.values().resize_({nnz}),
+            input.sizes());
+  }
+
+  if (input.layout() == kSparseCsr) {
+    static_cast<SparseCsrTensorImpl*>(input.unsafeGetTensorImpl())
+        ->set_member_tensors(
+            input.crow_indices(),
+            input.col_indices().resize_({nnz}),
+            input.values().resize_({nnz}),
+            input.sizes());
+  }
+}
+
+template <typename scalar_t, typename index_t>
+const Eigen::Map<Eigen::SparseMatrix<scalar_t, Eigen::ColMajor, index_t>>
+to_EigenCsc(const at::Tensor& tensor) {
+  int64_t rows = tensor.size(0);
+  int64_t cols = tensor.size(1);
+  int64_t nnz = tensor._nnz();
+
+  index_t* ccol_indices_ptr = tensor.ccol_indices().data_ptr<index_t>();
+  index_t* row_indices_ptr = tensor.row_indices().data_ptr<index_t>();
+  scalar_t* values_ptr = tensor.values().data_ptr<scalar_t>();
+  Eigen::Map<Eigen::SparseMatrix<scalar_t, Eigen::ColMajor, index_t>> map(
+      rows, cols, nnz, ccol_indices_ptr, row_indices_ptr, values_ptr);
+  return map;
+}
+
+template <typename scalar_t, typename index_t>
+const Eigen::Map<Eigen::SparseMatrix<scalar_t, Eigen::RowMajor, index_t>>
+to_EigenCsr(const at::Tensor& tensor) {
+  int64_t rows = tensor.size(0);
+  int64_t cols = tensor.size(1);
+  int64_t nnz = tensor._nnz();
+
+  index_t* crow_indices_ptr = tensor.crow_indices().data_ptr<index_t>();
+  index_t* col_indices_ptr = tensor.col_indices().data_ptr<index_t>();
+  scalar_t* values_ptr = tensor.values().data_ptr<scalar_t>();
+  Eigen::Map<Eigen::SparseMatrix<scalar_t, Eigen::RowMajor, index_t>> map(
+      rows, cols, nnz, crow_indices_ptr, col_indices_ptr, values_ptr);
+  return map;
+}
+
+template <typename scalar_t, typename index_t>
+void EigenCsr_to_tensor(
+    const at::Tensor& tensor,
+    const Eigen::SparseMatrix<scalar_t, Eigen::RowMajor, index_t>& matrix) {
+  TORCH_CHECK(
+      tensor.layout() == kSparseCsr,
+      "EigenCsr_to_tensor, expected tensor be kSparseCsr, but got",
+      tensor.layout());
+
+  int64_t nnz = matrix.nonZeros();
+  int64_t rows = matrix.outerSize();
+  csr_col_indices_and_values_resize(tensor, nnz);
+
+  if (nnz > 0) {
+    std::memcpy(
+        tensor.values().data_ptr<scalar_t>(),
```

Comment:
out of curiosity, how do we know that tensor.values() is big enough to hold all the values?

Reply:
This tensor is being resized just a couple of lines above:
    `csr_col_indices_and_values_resize(tensor, nnz);`

Thanks to your comment I looked up to its implementation and found out there was a duplicated code https://github.com/pytorch/pytorch/pull/101814/commits/99c1124f5e703f6bddb1a7376340126786c6b8a6  :smile: